container/rules/debian_pkg_tar.bzl

```diff
@@ -0,0 +1,187 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("@base_images_docker//package_managers:download_pkgs.bzl", _download_deb_pkgs = "download")
+load("@base_images_docker//package_managers:apt_key.bzl", _apt_key = "key")
+
+def _input_validation(kwargs):
+    allowed_attribues = ["name", "base", "language_layers"]
+
+    for key in kwargs.keys():
+        if key not in allowed_attribues:
+            fail("Attribute " + key + " is not supported.")
+
+container = [
+    ".tar.gz",
+    ".tgz",
+    ".tar",
+    ".tar.xz",
+]
+
+generate_deb_tar_attrs = _download_deb_pkgs.attrs + {
+    "base": attr.label(allow_files = container),
+    "packages": attr.string_list(),
+    "keys": attr.label_list(
+        allow_files = True,
+    ),
+}
+
+aggregate_debian_pkgs_attrs = {
+    "base": attr.label(allow_files = container),
+    "language_layers": attr.label_list(),
+
+    # Declare the following attributes since _download_deb_pkgs.implementation
+    # need access those attribute if their overrides are None
+    "additional_repos": attr.string_list(),
+    "_image_id_extractor": attr.label(
+        default = "@io_bazel_rules_docker//contrib:extract_image_id.py",
+        allow_files = True,
+        single_file = True,
+    ),
+}
+
+InstallableTarInfo = provider(fields = [
+    "installables_tar",
+])
+
+def _generate_deb_tar(
+        ctx,
+        packages = None,
+        additional_repos = None,
+        keys = None,
+        download_pkgs_output_tar = None,
+        download_pkgs_output_script = None):
+    """A function for producing a tarball for a set of debian packages.
+
+    Args:
+      ctx: ctx has either generate_deb_tar_attrs or aggregate_debian_pkgs_attrs
+      ctx depends on the rule that uses this function
+      packages: packages aggregated from each language_tool_layer by toolchain_container rule
+      additional_repos: additional_repos aggregated from each language_tool_layer by
+      toolchain_container rule
```

Comment:
these docs are too specific to the use of the rule from toolchain_container, when in fact, the rule gets called by language_tool_layer.
Just describe the attrs w/o reference to any of the rules that use them
e.g.
additional_repos: list of additional debian package repos to use,
          in sources.list format.
(see description of attrs in https://github.com/bazelbuild/bazel-toolchains/blob/master/container/rules/docker_toolchains.bzl#L342)

Reply:
Yea, you are right, thanks for giving the reference. Thanks.